Scenarios documentation (#49)
loadimpact/k6-docs: src/data/markdown/docs/01 guides/02 Using k6/14 Scenarios.md

This is an optional feature and existing scripts and options will continue to work as before
(with a few minor breaking changes, as mentioned
in the [release notes](https://github.com/loadimpact/k6/releases/tag/v0.27.0)),
but please [create a bug issue](https://github.com/loadimpact/k6/issues/new?labels=bug&template=bug_report.md)
if that's not the case.

There are several benefits of using scenarios in your tests:
- Multiple scenarios can be declared in the same script, and each one can
  execute a different JavaScript function, which makes organizing tests easier
  and more flexible.
- Every scenario can use a distinct VU and iteration scheduling pattern,
  powered by a purpose-built [executor](#executors). This enables modeling
  of advanced execution patterns which can better simulate real-world traffic.
- They can be configured to run in sequence or parallel.
- Different environment variables and tags can be set per scenario.

If you're excited to try it out, make sure you're using
[k6 v0.27.0 or above](https://github.com/loadimpact/k6/releases) and keep reading!
Note that scenarios are supported both in k6 CLI (local) and
[k6 Cloud](/cloud) execution.

## Configuration

> Scenarios are primarily configured via the `options` object in your test script. For example:

Reviewer: ```suggestion
Execution scenarios are primarily configured via `scenarios` key of the the exported `options` object in your test scripts. For example:
```

  }
}
```

</div>

### Common configuration

| Option         | Type   | Description                                                                      | Default     |
|----------------|--------|----------------------------------------------------------------------------------|-------------|
| `executor`     | string | Unique executor name. See the list of possible values below.                     | -           |
| `startTime`    | string | Time offset since test start this scenario should begin execution.               | `"0s"`      |
| `gracefulStop` | string | Time to wait for iterations to finish executing before stopping them forcefully. See the [gracefulStop](#graceful-stop-and-ramp-down) section. | `"30s"`     |
| `exec`         | string | Name of exported JS function to execute.                                         | `"default"` |
| `env`          | object | Environment variables specific to this scenario.                                 | `{}`        |
| `tags`         | object | [Tags](/using-k6/tags-and-groups) specific to this scenario.                     | `{}`        |

Possible values for `executor` are the executor name separated by hyphens:

> - `shared-iterations`

Reviewer: If possible, these should be links to the section anchors below

Author: Sure, good idea. Resolved by a57a3c1.

- `per-vu-iterations`
- `constant-vus`
- `ramping-vus`
- `constant-arrival-rate`
- `ramping-arrival-rate`
- `externally-controlled`

## Executors

> Executors are the workers of the k6 execution engine. Each one schedules VUs and

Reviewer: hmm "workers" here is a bit misleading - I consider the VUs to be the workers in the architecture... so, "workhorses" (metaphorical) or "worker schedulers" (factual) seem a bit better

```suggestion
Executors are the workhorses of the k6 execution engine. Each one schedules VUs and
```

Author: Yeah, that's more suitable :+1: Though I was looking at it from the perspective of the execution engine, i.e. executors do work for the engine. And VUs would be workers for executors, I guess? :smile: No matter, "workhorses" should paint the picture for everyone.

iterations differently, and you'll choose one depending on the type of traffic you
want to model to test your services.

### Shared iterations

A fixed number of iterations are "shared" between a number of VUs, and the test ends
once all iterations are executed. This executor is equivalent to the global `vus` and
`iterations` options.

Note that iterations aren't fairly distributed with this executor, and a VU that
executes faster will complete more iterations than others.

#### Options

| Option        | Type    | Description                                                                    | Default |
|---------------|---------|--------------------------------------------------------------------------------|---------|
| `vus`         | integer | Number of VUs to run concurrently.                                             | `1`     |
| `iterations`  | integer | Total number of script iterations to execute across all VUs.                   | `1`     |

> | `maxDuration` | string  | Maximum test duration before it's forcibly stopped (excluding `gracefulStop`). | `"10m"` |

Reviewer: ```suggestion
| `maxDuration` | string  | Maximum scenario duration before it's forcibly stopped (excluding `gracefulStop`). | `"10m"` |
```

#### Examples

- Execute 200 total iterations shared by 10 VUs with a maximum duration of 10 seconds:

<div class="code-group" data-props='{"labels": [ "shared-iters.js" ], "lineNumbers": "[true]"}'>

```js
import http from 'k6/http';

export let options = {
  discardResponseBodies: true,
  scenarios: {
    contacts: {
      executor: 'shared-iterations',
      vus: 10,
      iterations: 200,
      maxDuration: '10s',
    }
  }
};

export default function() {
  http.get('https://test.k6.io/contacts.php');
}
```

</div>

### Per VU iterations

Each VU executes an exact number of iterations. The total number of completed
iterations will be `vus * iterations`.

#### Options

| Option        | Type    | Description                                                                    | Default |
|---------------|---------|--------------------------------------------------------------------------------|---------|
| `vus`         | integer | Number of VUs to run concurrently.                                             | `1`     |
| `iterations`  | integer | Number of script iterations to execute with each VU.                           | `1`     |

> | `maxDuration` | string  | Maximum test duration before it's forcibly stopped (excluding `gracefulStop`). | `"10m"` |

Reviewer: ```suggestion
| `maxDuration` | string  | Maximum scenario duration before it's forcibly stopped (excluding `gracefulStop`). | `"10m"` |
```

#### When to use

> Use this executor if you need a specific amount of VUs to complete the same amount of
> iterations.

Reviewer: ```suggestion
Use this executor if you need a specific amount of VUs to complete the same amount of
iterations. This can be useful when you have fixed sets of test data that you want to 
partition between VUs.
```

#### Examples

- Execute 20 iterations by 10 VUs *each*, for a total of 200 iterations, with a

>   maximum duration of 10 seconds:

Reviewer: doesn't really matter, but I wanted to give more examples of time units
```suggestion
  maximum duration of 1 hour and 30 minutes:
```

<div class="code-group" data-props='{"labels": [ "per-vu-iters.js" ], "lineNumbers": "[true]"}'>

```js
import http from 'k6/http';

export let options = {
  discardResponseBodies: true,
  scenarios: {
    contacts: {
      executor: 'per-vu-iterations',
      vus: 10,
      iterations: 20,

>       maxDuration: '10s',

Reviewer: ```suggestion
      maxDuration: '1h30m',
```

    }
  }
};

export default function() {
  http.get('https://test.k6.io/contacts.php');
}
```

</div>

### Constant VUs

A fixed number of VUs execute as many iterations as possible for a specified amount
of time. This executor is equivalent to the global `vus` and `duration` options.

#### Options

| Option     | Type    | Description                                     | Default |
|------------|---------|-------------------------------------------------|---------|
| `vus`      | integer | Number of VUs to run concurrently.              | `1`     |

> | `duration` | string  | Total test duration (excluding `gracefulStop`). | -       |

Reviewer: ```suggestion
| `duration` | string  | Total scenario duration (excluding `gracefulStop`). | -       |
```

#### When to use

Use this executor if you need a specific amount of VUs to run for a certain amount of time.

#### Examples

> - Run a constant 10 VUs for 10 seconds:

Reviewer: ```suggestion
- Run a constant 10 VUs for 45 minutes:
```

<div class="code-group" data-props='{"labels": [ "constant-vus.js" ], "lineNumbers": "[true]"}'>

```js
import http from 'k6/http';

export let options = {
  discardResponseBodies: true,
  scenarios: {

>     contacts: {

Reviewer: ```suggestion
    my_awesome_api_test: {
```

      executor: 'constant-vus',
      vus: 10,

>       duration: '10s',

Reviewer: ```suggestion
      duration: '45m',
```

    }
  }
};

export default function() {

>   http.get('https://test.k6.io/contacts.php');

Reviewer: just adding some diversity to the examples
```suggestion
  http.get('https://test-api.k6.io/');
  sleep(Math.random() * 3);
```

#### When to use

This executor is a good fit if you need VUs to ramp up or down during specific periods
of time.

#### Examples

- Run a two-stage test, ramping up from 0 to 100 VUs for 5 seconds, and down to 0 VUs
  for 5 seconds:

<div class="code-group" data-props='{"labels": [ "ramping-vus.js" ], "lineNumbers": "[true]"}'>

```js
import http from 'k6/http';

export let options = {
  discardResponseBodies: true,
  scenarios: {
    contacts: {
      executor: 'ramping-vus',
      startVUs: 0,
      stages: [
        { duration: '5s', target: 100 },
        { duration: '5s', target: 0 },
      ],
      gracefulRampDown: '0s',
    }
  }
};

export default function() {
  http.get('https://test.k6.io/contacts.php');
}
```

</div>

Note the setting of `gracefulRampDown` to 0 seconds, which could cause some
iterations to be interrupted during the ramp down stage.

### Constant arrival rate

A fixed number of iterations are executed in a specified period of time.
Since iteration execution time can vary because of test logic or the
system-under-test responding more slowly, this executor will try to compensate

> by running a variable number of VUs&mdash;including initializing more in the middle

Reviewer: ```suggestion
by running a variable number of VUs&mdash;including potentially initializing more in the middle
```

of the test&mdash;in order to meet the configured iteration rate. This approach is
useful for a more accurate representation of RPS, for example.

See the [arrival rate](#arrival-rate) section for details.

#### Options

| Option            | Type    | Description                                                                             | Default |
|-------------------|---------|-----------------------------------------------------------------------------------------|---------|
| `rate`            | integer | Number of iterations to execute each `timeUnit` period.                                 | -       |
| `timeUnit`        | string  | Period of time to apply the `rate` value.                                               | `"1s"`  |

> | `duration`        | string  | Maximum test duration before it's forcibly stopped (excluding `gracefulStop`).          | -       |

Reviewer: ```suggestion
| `duration`        | string  | Total scenario duration (excluding `gracefulStop`).                                     | -       |
```

| `preAllocatedVUs` | integer | Number of VUs to pre-allocate before test start in order to preserve runtime resources. | -       |
| `maxVUs`          | integer | Maximum number of VUs to allow during the test run.                                     | -       |

#### When to use

When you want to maintain a constant number of requests without being affected by the

> system-under-test's performance.

Reviewer: ```suggestion
performance of the system under test.
```

      preAllocatedVUs: 50,
      maxVUs: 100,
      stages: [
        { target: 200, duration: '30s' },
        { target: 0, duration: '30s' },
      ],
    }
  }
};

export default function() {
  http.get('https://test.k6.io/contacts.php');
}
```

</div>

### Externally controlled

Control and scale execution at runtime via [k6's REST API](/misc/k6-rest-api) or
the [CLI](https://k6.io/blog/how-to-control-a-live-k6-test).

This executor formalizes what was previously possible globally, and using it is
required in order to use the `pause`, `resume`, and `scale` CLI commands. Also note
that arguments to `scale` to change the amount of active or maximum VUs only affect
the externally controlled executor.

#### Options

| Option     | Type    | Description                                         | Default |
|------------|---------|-----------------------------------------------------|---------|
| `vus`      | integer | Number of VUs to run concurrently.                  | -       |
| `maxVUs`   | integer | Maximum number of VUs to allow during the test run. | -       |
| `duration` | string  | Total test duration.                                | -       |

#### When to use

> If you want to control the number of VUs while the test is running.

Reviewer: ```suggestion
If you want to control the number of VUs while the test is running.

Important: this is the only executor that is not supported in `k6 cloud`, it can only be used locally with `k6 run`.
```

In other words, when the target system is being stressed and starts to respond more
slowly a closed model load test will play "nice" and wait, resulting in increased
iteration durations and a tapering off of the arrival rate of new VU iterations.

This is not ideal when the goal is to simulate a certain arrival rate of new VUs,
or more generally throughput (eg. requests per second).

To fix this problem we use an open model, decoupling the start of new VU iterations
from the iteration duration and the influence of the target system's response time.

![Arrival rate closed/open models](images/arrival-rate-open-closed-model.png)

In k6 we've implemented this open model with our "arrival rate" executors. There are
two arrival rate executors to chose from for your scenario(s),
[constant-arrival-rate](#constant-arrival-rate) and [ramping-arrival-rate](#ramping-arrival-rate):

<div class="code-group" data-props='{"labels": [ "open-model.js" ], "lineNumbers": "[true]"}'>

```js
import http from 'k6/http';

export let options = {
  scenarios: {
    open_model: {
      executor: 'constant-arrival-rate',
      rate: 1,
      timeUnit: '1s',
      duration: '1m',

>       preAllocatedVUs: 1,
>       maxVUs: 100,

Reviewer: Allocating VUs in the middle of the test is probably a bad practice in general. While we support it, we shouldn't give example where it's obvious that the preAllocatedVUs will be very insufficient.
```suggestion
      preAllocatedVUs: 20,
      maxVUs: 100,
```
